java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/debugging/launch/NbLaunchDelegate.java

```diff
@@ -200,30 +201,43 @@ public void progressHandleCreated(ProgressOperationEvent e) {
             List<String> args = argsToStringList(launchArguments.get("args"));
             if (debug) {
                 requestProcessor.post(() -> {
-                    Lookups.executeWith(launchCtx, () -> {
+                    ActionProgress debugProgress = ActionProgress.start(launchCtx);
+                    // Add session's lookup, it may override dialog displayer, etc.
+                    Lookups.executeWith(new ProxyLookup(launchCtx, context.getLspSession().getLookup()), () -> {
                         String miDebugger = (String) launchArguments.get("miDebugger");
-                        startNativeDebug(nativeImageFile, args, miDebugger, context, executionDescriptor, launchFuture);
+                        startNativeDebug(nativeImageFile, args, miDebugger, context, executionDescriptor, launchFuture, debugProgress);
                     });
                 });
             } else {
+                AtomicReference<Future<Integer>> run = new AtomicReference<>();
                 Lookups.executeWith(launchCtx, () -> {
-                    execNative(nativeImageFile, args, context, executionDescriptor, launchFuture);//, success);
+                    run.set(execNative(nativeImageFile, args, context, executionDescriptor, launchFuture));
+                });
+                new RequestProcessor().post(() -> {
+                    Integer ret;
```

Comment:
This looks like the old, ugly pattern. Since https://github.com/apache/netbeans/pull/2938/files#r635969684 once can get the `exitCode` (or `null`) synchronously as soon as execution is over.

Reply:
That significantly simplified the code, thanks.
And it uncovered a bug in cpplite debugger - the GDB process did not finish after application exit. This is fixed, including a test, as well.